### PR TITLE
Add batch span processor benchmarks

### PR DESCRIPTION
**Pull request by sbandadd**
Description:
This PR adds two benchmarks.

1. Current benchmark executs forceFlush() on every loop and creates a bottleneck which results in not stressing batch span processor. Current benchmark only
measures throughput which is not helpful on its own since number of spans getting exported is also important. BatchSpanProcessorMultiThreadBenchmark is created to address this issue.
2. Measuring CPU usage of exporter thread is also important, but the current benchmarks consumes as much CPU as possible which makes the measurement not meaningful.
To maintain a steady state, this PR creates a benchmark that generates 10k spans per second per thread. One would need to attach a profiler such as yourkit or JProfiler
to the benchmark run to understand the processor's CPU usage. BatchSpanProcessorCpuBenchmark is created for this purpose.

This PR also fixes a bug in calculating number of dropped/export spans. Earlier dropped and exported spans were actually counting the other one.

### CODE REVIEW

File: sdk/trace/src/jmh/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessorCpuBenchmark.java
Comment (anuraaga) [inline sdk/trace/src/jmh/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessorCpuBenchmark.java:94]: Thanks for helping with benchmarks - how is the number decided here? 

For overhead benchmarks, I usually avoid jmh because of this but if there's a way to simulate fixed load with jmh that's pretty cool.

File: sdk/trace/src/jmh/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessorMetrics.java
Comment (anuraaga) [inline sdk/trace/src/jmh/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessorMetrics.java:48]: mapToLong

File: sdk/trace/src/jmh/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessorMetrics.java
Comment (anuraaga) [inline sdk/trace/src/jmh/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessorMetrics.java:41]: I have a feeling two loops will be very similar code while more readable

File: sdk/trace/src/jmh/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessorMultiThreadBenchmark.java
Comment (anuraaga) [inline sdk/trace/src/jmh/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessorMultiThreadBenchmark.java:49]: You call register global do don't need this

File: sdk/trace/src/jmh/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessorCpuBenchmark.java
Comment (sbandadd) [reply sdk/trace/src/jmh/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessorCpuBenchmark.java:94]: This is hand picked. 10k sounds reasonable to me for testing

File: sdk/trace/src/jmh/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessorMultiThreadBenchmark.java
Comment (sbandadd) [reply sdk/trace/src/jmh/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessorMultiThreadBenchmark.java:49]: Nice. Will remove.

File: sdk/trace/src/jmh/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessorMetrics.java
Comment (sbandadd) [reply sdk/trace/src/jmh/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessorMetrics.java:41]: The existing code is two loop and it took me a good amount of time to debug why dropped/exported spans metric are invalid and got a bit confused on what the two loops are doing. Functional would make it easy to see what are getting filtered and how the data is getting mapped.

That said, I would leave this to the maintainers.

File: sdk/trace/src/jmh/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessorCpuBenchmark.java
Comment (anuraaga) [reply sdk/trace/src/jmh/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessorCpuBenchmark.java:92]: We need to know how it was hand picked, especially since these will need to be maintained by other folks, possibly adding / tweaking benchmarks. Is it just a large amount of time to dwarf the time spent on the rest of the work?

Comment (sbandadd) [reply sdk/trace/src/jmh/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessorCpuBenchmark.java:92]: Yes, it is a large amount of time. As shown in my runs, this clearly surfaced CPU overhead.

# Chart
image

Comment (anuraaga) [reply sdk/trace/src/jmh/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessorCpuBenchmark.java:92]: So let's leave a comment that the time spent in the rest of the work is almost nothing, allowing this sleep to control the rate span generation.

Comment (sbandadd) [reply sdk/trace/src/jmh/java/io/opentelemetry/sdk/trace/export/BatchSpanProcessorCpuBenchmark.java:92]: Added

**Review by anuraaga — APPROVED** [review]
Having fixed-throughput overhead benchmarks is very helpful, thanks!

**Comment by jkwatson** [comment]
I finally got to spend some dedicated time on this today, and got jmh hooked up with the async-profiler profiling option, and I do see some significant overhead introduced by the usage of the ArrayBlockingQueue (although it's been tough to get really consistent results).

thanks for putting in the work on this. Now, to agree on the best solution to the issue. :)

**Comment by anuraaga** [comment]
@jkwatson Approved of these benchmarks offline so will go ahead and merge. Benchmarks are alive so I'm sure we'll continue to tweak but this is an improvement. Thanks @sbandadd

**Pull request merged**
